drivers/usb-common.c

```diff
@@ -410,3 +410,80 @@ void warn_if_bad_usb_port_filename(const char *fn) {
 		__func__, fn);
 	return;
 }
+
+/* API neutral, handles retries */
+static int nut_usb_get_string_descriptor(
+	usb_dev_handle *udev,
+	int StringIdx,
+	int langid,
+	char *buf,
+	size_t buflen)
+{
+	int ret = -1;
+	int retries = 3;
+
+	while (retries--) {
+		ret = usb_get_string(udev, (usb_ctrl_strindex)StringIdx, langid, (usb_ctrl_charbuf)buf, buflen);
+		if (ret >= 0) {
+			break;
+		} else if (retries) {
+			upsdebugx(1, "%s: string descriptor %d request failed, retrying...", __func__, StringIdx);
+			usleep(50);
+		}
+	}
+	return ret;
+}
+
+/* API neutral, assumes en_US if langid descriptor is broken */
+int nut_usb_get_string(
+	usb_dev_handle *udev,
+	int StringIdx,
+	char *buf,
+	size_t buflen)
+{
+	int ret;
+	char buffer[255];
+	int langid;
+	int len;
+	int i;
+
+	if (!udev || StringIdx < 1 || StringIdx > 255) {
+		return -1;
+	}
+
+	/* request langid descriptor */
+	ret = nut_usb_get_string_descriptor(udev, 0, 0, buffer, 4);
+	if (ret < 0)
+		return ret;
+
+	if (ret == 4 && buffer[0] >= 4 && buffer[1] == USB_DT_STRING) {
+		langid = buffer[2] | (buffer[3] << 8);
+	} else {
+		upsdebugx(1, "%s: Broken language identifier, assuming en_US", __func__);
+		langid = 0x0409;
+	}
+
+	/* retrieve string in preferred language */
+	ret = nut_usb_get_string_descriptor(udev, StringIdx, langid, buffer, sizeof(buffer));
+	if (ret < 0) {
```

Comment:
Yes. Unless you know about a device that:
- returns a valid langid &&
- the valid reported langid is not en_US &&  #already here it is gonna be difficult to find examples
- doesn't work with the reported langid &&  #would be an "interesting" bug, would probably fail on most OS'es
- works with en_US

Reply:
Ok, sounds fair :)

Reply:
"In general we shouldn't add extra code "just in case, cannot hurt" unless it is has been proven to help at least in one real case." applies here as well :)